core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1745,11 +1745,16 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * has overridden the call site using `setCallSite()`, this will return the user's version.
    */
   private[spark] def getCallSite(): CallSite = {
-    val callSite = Utils.getCallSite()
```

Comment:
I thought I tried that and it didn't work, but I must have done something wrong. It seems to:

```
scala> def foo(): Int = { println("foo"); 42 }
foo: ()Int

scala> def bar(arg: Boolean): Int = { lazy val f = foo(); if (arg) { f } else { 0 } }
bar: (arg: Boolean)Int

scala> bar(true)
foo
res0: Int = 42

scala> bar(false)
res1: Int = 0
```

So yeah that could be a much cleaner solution.